---
title: "SingleDataset"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{SingleDataset}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# This vignette will provide guidance for analysis of a single dataset.


#### package installation:
```
install.packages("devtools") #if devtools not already installed
devtools::install_github('apf2139/tamlabscpipeline')
```
#### dependency installation:
```
install.packages("tidyverse")
install.packages('Seurat')
devtools::install_github(repo = 'ChristophH/sctransform')
install.packages("cowplot")
devtools::install_github('chris-mcginnis-ucsf/DoubletFinder')
```

### startup

Let's assume you have a dataset at the ready. I will try to provide a small example dataset for reproducible tutorials in the future.

First, load up the package and dependencies. running `library(tamlabscpipeline)` will examine the depencies and compare your versions with the latest known compatible versions for `tamlabscpipeline` and issue warnings if you are behind / ahead of the latest test.



```
library(Seurat)
library(sctransform)
library(tidyverse)
library(cowplot)
library(DoubletFinder)
```

```{r setup}

library(tamlabscpipeline)
```


Let us assume also that you have access to an output of cellranger in the form of a dir or an H5 file. 

The seuratpipeline() function will perform QC and initial clustering of your dataset. This will involve filtering out high mitochondrial content cells, low library size cells, and doublets; then normalizing using [SingleCellTransform (SCT)](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1) and clustering using the Louvain method as implemented in [Seurat](https://satijalab.org/seurat/).

Here is a quick start for launching the analysis pipeline using this function. I have used the object filepath to store a string containing the path to an example file.

The command prints out a bunch of plots for QC purposes, so best to wrap it in a pdf command block.

### quick start

```
pdf('testrun.pdf')

sobj <- seuratpipeline(data = filepath, #path to dataset
               format = h5 #format; dir or H5
               
               )

dev.off()

```

This function is designed to give a rough but pretty good minimum initial clustering across different datasets. On a dataset of ~17,000 genes x ~5,500 using a 2018 MacBook Pro this took roughly 20 minutes. In that time, you can read the rest of this vignette for more details about this function and how to further assess the quality of your dataset; think about some hypotheses you'd like to explore in your data;  and ponder the precarious global geopolitical situation we find ourselves living in.

***

By default, this function will perform QC and clustering of the data. Default QC steps are as follows:

1. First-pass clustering involving normalization via SCT and Louvain clustering in Seurat with default parameters.
2. Identification and removal of mitochondrial cluster. Without any prefiltering, there will almost certainly be a cluster strongly driven by mitochondrial RNA content.
  + this step will generate two violin plots; one before removal and one after removal and reclustering.
3. Second-pass clustering after mito cluster removal. Iterative library size filtering.
  + By default, the second-pass clustering step now only involves clustering based on minimum library size. Proper thesholds for minimum library size are based on median absolute deviation thresholds which are learnt directly from the data.
  + Filtering will not take place if the cluster size is below a minimum arbitrary threshold (100 cells), as Med.Abs.Dev. learning performs quite poorly with smaller cell numbers.
  + this step will generate two visualizations for each cluster: one showing the Med.Abs.Dev. threshold learnt for that cluster, the second showing the actual cutoff point denoted by this threshold. Plots will be generated even for clusters of under 100 cells, but these clusters will not actually be subjected to lib size filtration.
4. Third-pass clustering. This will involve the identification and removal of doublets. Doublet identification relies on [DoubletFinder](https://www.sciencedirect.com/science/article/abs/pii/S2405471219300730)
  + this will generate a line-graph containing very little information that you can and should ignore. This step also performs its own multi-pass clustering steps using default parameters.
5. Cell cycle scores will be calculated. By default, there will be no automatic cell cycle correction, as this can vary from dataset to dataset and should be assessed manually.
6. A finalized normalization and clustering step using SCT and Seurat Louvain clustering will take place.
7. The function will return a Seurat object containing fully normalized and scaled matrices; clustering assignments; and dimensionality reduction embeddings, including PCA, tSNE and UMAP.



#### Here are some suggestions for assessing dataset quality.

``` {r, echo=FALSE}
library(Seurat)
library(sctransform)
library(tidyverse)
library(cowplot)
library(DoubletFinder)
knitr::opts_chunk$set(fig.width=7, fig.height=5) 
sobj <- readRDS('../../tamlabtesting/sobj_default.rds')
```

``` {r}

FeaturePlot(sobj, c('nCount_RNA', 'percent.mito', 'S.Score', 'G2M.Score'))

VlnPlot(sobj, c('nCount_RNA', 'nFeature_RNA', 'percent.mito'), ncol = 1)

DimPlot(sobj, group.by = 'Phase')

```

A full assesment of quality will require cell type identification, for example via specific cell type markers. An example of this using Seurat's `FeaturePlot()` function is shown below:

``` {r}

FeaturePlot(sobj, c('Ptprc', 'Cd3e', 'Cd19', 'Epcam'))

```

Other cell type calling approaches make use of differentially expressed genes, the calling of which is discussed below. Automated (at at least semi-automated) cell-type calling is an active area of development and will be included in this suite soon.

An integrated assessment of quality and the decision to recluster using different QC parameters is thus a non-trivial and manual process, but these functions and the different parameters in the `seuratpipeline()` function will allow easy tweaking.

***

## Downstream analysis

After clustering, it is often useful to run differential expression testing to identify differentiall expressed genes (DEGs). This can be used as input for Gene Set Enrichment Analysis (GSEA).

For differential expression testing, Seurat provides wrappers for tests such as the [MAST](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4676162/) test, specifically developed to identify DEGs while attempting to correct for the biases and noise inherent to scRNAseq data.

The following can be used to test for DEGs in all clusters:

```
library(MAST)

sobj.markers <- FindAllMarkers(sobj, 
                               test.use = 'MAST',
                               latent.vars = c('nCount_RNA', 'percent.mito')
```

Exploration of these can occur manually via manipulation of the resulting data.frame structure.


GSEA functions are provided by the `tamlabscpipeline` suite. Documentation and vignettes will be provided very soon.

***

There are some major improvements over the previous versions, including much better cell QC (more accrurately removing low-quality cells while being nicer to okay-quality cells). Also, there have been important changes made to clean up the code, remove unnecesarry options, optimize speed, and retain full scaled.data matrices at the final output.


``` {r, echo=FALSE}
rm(sobj)
```
